katcp/testutils.py

```diff
@@ -1809,6 +1810,39 @@ def assert_wait_call_count(self, count, timeout=1.):
         if self.call_count >= count and len(self.call_args_list) < count:
             raise RuntimeError("call_args_list not updated within timeout.")
 
+class AsyncWaitingMock(mock.Mock):
+
+    def __init__(self, *args, **kwargs):
+        super(AsyncWaitingMock, self).__init__(*args, **kwargs)
+        self._call_event = tornado.locks.Event()
+
+    def _mock_call(self, *args, **kwargs):
+        ret = super(AsyncWaitingMock, self)._mock_call(*args, **kwargs)
+        self._call_event.set()
+        return ret
+
+    @tornado.gen.coroutine
+    def assert_wait_call_count(self, count, timeout=1.):
+        """Wait for mock to be called >= *count* times within *timeout* seconds.
+
+        Raises AssertionError if the call count is not reached.
+
+        """
+        t0 = time.time()
+        if self.call_count >= count and len(self.call_args_list) >= count:
+            raise tornado.gen.Return(True)
+        to_wait = timeout
+
+        while to_wait >= 0 and self.call_count < count:
+            try:
+                yield self._call_event.wait(timeout)
+                self._call_event.clear()
+            except gen.TimeoutError:
```

Comment:
Slightly inconsistent use of the tornado namespace: lines 1824 and 1833 use `tornado.gen`, while here we just use `gen` directly.  Interestingly, we don't have an `import tornado` or `import tornado.gen` line in this file, so is another import bringing `tornado.gen` into the namespace, or do I just misunderstand these things?

On a probably unrelated note, I see the Jenkins console output reports a few of these immediately after running the tests.
```
Exception AttributeError: "'NoneType' object has no attribute 'gen'" in <generator object connect at 0x7f75a490fd20> ignored
```
Any idea what that is?